File: kotlin/divide&conquer/main_6549.kt
import kotlin.math.max
import kotlin.math.min

lateinit var array: IntArray

fun main() {
    while (true) {
        array = readLine()!!.split(' ').map { it.toInt() }.toIntArray()
        if (array[0] == 0)
            return
        println(cut(1, array[0]))
    }
}

fun cut(start: Int, end: Int): Long {
    if (start == end) {
        return array[start].toLong()
    }

    val mid = (start + end) / 2
    var result = max(cut(start, mid), cut(mid + 1, end))

    var low = mid
    var high = mid + 1
    var height = min(array[low], array[high]).toLong()
    result = max(result, height * 2)

    while (start < low || high < end) {
        height = when {
            low <= start -> {
                min(height, array[++high].toLong())
            }
            end <= high -> {
                min(height, array[--low].toLong())
            }
            array[low - 1] > array[high + 1] -> {
                min(height, array[--low].toLong())
            }
            else -> {
                min(height, array[++high].toLong())
            }
        }
        result = max(result, height * (high - low + 1))
    }
    return result
}

File: cpp/divide&conquer/main_6549.cpp
#include <iostream>
#include <cmath>
using namespace std;

long long area;
int arr[100000];

void find(const int start, const int end) {
	if (start > end) {
		return;
	}
	if (start == end) {
		area = max(area, static_cast<long long>(arr[start]));
		return;
	}

	const int mid = (start + end) / 2;
	find(start, mid);
	find(mid + 1, end);


	int low = mid;
	int high = mid + 1;
	int height = min(arr[low], arr[high]);
	area = max(area, static_cast<long long>(height) * 2);
	while (start < low && high < end) {
		if (low <= start) {
			height = min(height, arr[++high]);
		} else if (high >= end) {
			height = min(height, arr[--low]);
		} else if (arr[high + 1] > arr[low - 1]) {
			height = min(height, arr[++high]);
		} else {
			height = min(height, arr[--low]);
		}
		area = max(area, height * (static_cast<long long>(high) - low + 1));
	}
}

int main() {
	ios_base::sync_with_stdio(false);
	cin.tie(nullptr);
	cout.tie(nullptr);

	int num;
	while (true) {
		area = 0;
		cin >> num;
		if (num == 0) {
			return 0;
		}
		for (int i = 0; i < num; ++i) {
			cin >> arr[i];
		}
		find(0, num - 1);
		cout << area << '\n';
	}
}
